### ISSUE TITLE
Implementing conduction delays ?

### ISSUE THREAD

**Issue by PaulMuadDib** [Hananel-Hazan/bindsnet#163]
Implementing a conduction delay for a connection, or a set of conduction delays to model directly an impulse response ? (for a connection)

**Comment by djsaunde** [comment]
Could you elaborate? PRs are welcome.

**Comment by PaulMuadDib** [comment]
For the time being, a connection is a weight, and I was wondering about the possibility to add a "conduction delay" attribute (say from 0 to 50 ms) to simulate the time needed for the action potential / spike to propagate from one neuron to the following... so that a connection would be a (weight , delay) in this manner: https://www.izhikevich.org/publications/spnet.cpp
I was also wondering about the possibility to extend this unique couple to a set of (weights , delays)  (effectively modelling an inpulse response).

**Comment by djsaunde** [comment]
That is an interesting idea, but I'm not sure we need it in BindsNET, especially not in the generic `Connection` object where it would induce computational overhead. Maybe implementing a new `DelayConnection` object with this functionality would be the way to go.

Personally, I don't have a use case for simulating propagation times, nor do I have the extra time to work on it, so I'd encourage you to make a PR if you'd like the functionality in the package.

**Comment by PaulMuadDib** [comment]
I understand, it is not my priority right now, and it would certainly slow down the computations. I may try to implement it as a new connection object later, but I am rather struggling to backpropagate rewards in an efficient manner for now.

**Comment by PaulMuadDib** [comment]
Hi there,
It appears that minor modifications on the "Connection" object (see DelayedConnection below) would allow the introduction of synaptic conduction delays (which, imo, are fundamentally required to represent/augment the adaptive spatiotemporal processing capabilities of a spiking neuron ruled by STDP, and allow the emergence of richer internal representations in the form of polychronous groups, see Izhikevich). I tried to optimize/vectorize the .compute( ) part (summation of the delayed weights * spikes contributions), but it still slows down the simulation by 4 with my deep SNN (with synaptic delays from 0 to 20 ms).

    `class AbstractDelayedConnection(ABC):
    """
    Abstract base method for connections between ``Nodes``.
    """

    def __init__(self, source: Nodes, target: Nodes,
                 nu: Optional[Union[float, Sequence[float]]] = None, weight_decay: float = 0.0, **kwargs) -> None:
        """
        Constructor for abstract base class for connection objects.

        :param source: A layer of nodes from which the connection originates.
        :param target: A layer of nodes to which the connection connects.
        :param nu: Learning rate for both pre- and post-synaptic events.
        :param weight_decay: Constant multiple to decay weights by on each iteration.

        Keyword arguments:

        :param function update_rule: Modifies connection parameters according to some rule.
        :param float wmin: The minimum value on the connection weights.
        :param float wmax: The maximum value on the connection weights.
        :param float norm: Total weight per target neuron normalization.
        """
        self.w = None
        self.d = None
        self.source = source
        self.target = target
        self.nu = nu
        self.weight_decay = weight_decay

        assert isinstance(source, Nodes), 'Source is not a Nodes object'
        assert isinstance(target, Nodes), 'Target is not a Nodes object'

        from ..learning import NoOp

        self.update_rule = kwargs.get('update_rule', NoOp)
        self.wmin = kwargs.get('wmin', None)
        self.wmax = kwargs.get('wmax', None)
        self.dmin = kwargs.get('dmin', None)
        self.dmax = kwargs.get('dmax', None)
        self.norm = kwargs.get('norm', None)
        self.decay = kwargs.get('decay', None)

        if self.update_rule is None:
            self.update_rule = NoOp

        self.a_pre = 0.0

        self.update_rule = self.update_rule(
            connection=self, nu=self.nu, weight_decay=weight_decay
        )

    @abstractmethod
    def compute(self, s: torch.Tensor) -> None:
        """
        Compute pre-activations of downstream neurons given spikes of upstream neurons.

        :param s: Incoming spikes.
        """
        pass

    @abstractmethod
    def update(self, **kwargs) -> None:
        """
        Compute connection's update rule.
        """
        learning = kwargs.get('learning', True)

        if learning:
            self.update_rule.update(**kwargs)

        mask = kwargs.get('mask', None)
        if mask is not None:
            self.w.masked_fill_(mask, 0)

    @abstractmethod
    def normalize(self) -> None:
        """
        Normalize weights so each target neuron has sum of incoming connection weights equal to ``self.norm``.
        """
        pass

    @abstractmethod
    def reset_(self) -> None:
        # language=rst
        """
        Contains resetting logic for the connection.
        """
        pass




    class DelayedConnection(AbstractDelayedConnection):
    """
    Specifies delayed synapses between one or two populations of neurons.
    """
    def __init__(self, source: Nodes, target: Nodes, nu: Optional[Union[float, Sequence[float]]] = None,
                 weight_decay: float = 0.0, **kwargs) -> None:
        """
        Instantiates a :code:`Connection` object.

        :param source: A layer of nodes from which the connection originates.
        :param target: A layer of nodes to which the connection connects.
        :param nu: Learning rate for both pre- and post-synaptic events.
        :param weight_decay: Constant multiple to decay weights by on each iteration.

        Keyword arguments:

        :param function update_rule: Modifies connection parameters according to some rule.
        :param torch.Tensor w: Strengths of synapses.
        :param torch.Tensor b: Target population bias.
        :param torch.Tensor d: Conduction delays of synapses.
        :param float wmin: Minimum allowed value on the connection weights.
        :param float wmax: Maximum allowed value on the connection weights.
        :param float dmin: Minimum allowed value on the connection weights.
        :param float dmax: Maximum allowed value on the connection weights.
        :param float norm: Total weight per target neuron normalization constant.
        """
        super().__init__(source, target, nu, weight_decay, **kwargs)

        self.w = kwargs.get('w', None)
        self.d = kwargs.get('d', None)

        if self.w is None:
            if self.wmin is None or self.wmax is None:
                self.w = torch.rand(source.n, target.n)
            elif self.wmin is not None and self.wmax is not None:
                self.w = self.wmin + torch.rand(source.n, target.n) * (self.wmax - self.wmin)
        else:
            if self.wmin is not None and self.wmax is not None:
                self.w = torch.clamp(self.w, self.wmin, self.wmax)

        self.b = kwargs.get('b', torch.zeros(target.n))

        if self.d is None:
            if self.dmin is None or self.dmax is None:
                self.d = torch.randint(low = 0,         high = source.d,  size=(source.n, target.n))
            elif self.dmin is not None and self.dmax is not None:
                self.d = torch.randint(low = self.dmin, high = self.dmax, size=(source.n, target.n)) 
        else:
            if self.dmin is not None and self.dmax is not None:
                self.d = torch.clamp(self.d, self.dmin, self.dmax)
				
		# To boost (=max vectorization) the .compute part
        self.nb_synapses   =  torch.zeros((self.dmax, 1), dtype = torch.int)
        self.delay_synapes =  torch.zeros((self.dmax, source.n * target.n // 3, 2), dtype = torch.int)
        for delay in range(self.dmax):
            connections             = (self.d == delay).nonzero()
            self.nb_synapses[delay] = connections.size(0)
            if connections.size(0) > 0:
                self.delay_synapes[delay, 0:connections.size(0)-1, 0:connections.size(1)-1] = connections[0:connections.size(0)-1, 0:connections.size(1)-1]
				
    def compute(self, s: torch.Tensor) -> torch.Tensor:
        """
        Compute pre-activations given source spikes using delayed / weighted connections

        :param s: Incoming spikes.
        :return: Incoming spikes multiplied by synaptic weights, biased and delayed (with or without decaying spike activation).
        """
        for delay in range(self.target.d):																			
            if self.nb_synapses[delay] > 0:																			# For each possible delay d of a connection
                presyn  = self.delay_synapes[delay,0:self.nb_synapses[delay]-1,0].long().view(-1)					# Presynaptic/source neurons of connections of delay d
                postsyn = self.delay_synapes[delay,0:self.nb_synapses[delay]-1,1].long().view(-1)					# Postsynaptic/target neurons of connections of delay d
                self.target.inpt[postsyn, delay] += self.w[presyn, postsyn].view(-1) @ s[presyn].float().view(-1)	# Add the d - delayed contribution w * s of each connections of delay d		
            if delay == 0:    post = self.target.inpt[:, delay].view(*self.target.shape)							# Current (0 delay) pre-activation to be returned  														
            else:             self.target.inpt[:, delay-1] = self.target.inpt[:, delay]								# Action potential propagation (simple time shift @ each call)
        self.target.inpt[:, self.target.d - 1] = 0.0
        return post																									

    def update(self, **kwargs) -> None:
        """
        Compute connection's update rule.
        """
        super().update(**kwargs)

    def normalize(self) -> None:
        """
        Normalize weights so each target neuron has sum of connection weights equal to ``self.norm``.
        """
        if self.norm is not None:
            self.w *= self.norm / self.w.abs().sum(0).unsqueeze(0)

    def reset_(self) -> None:
        """
        Contains resetting logic for the connection.
        """
        super().reset_()`

**Comment by Enguerran971** [comment]
Hi,
I've read your code and it's very interesting for me. I think you also modify the Nodes class (self.target.inpt and self.target.d), could you show me this class.
Thank's
